audit/hashstructure.go

```diff
@@ -108,18 +107,18 @@ func Hash(salter *salt.Salt, raw interface{}, nonHMACDataKeys []string) error {
 // the structure. Only _values_ are hashed: keys of objects are not.
 //
 // For the HashCallback, see the built-in HashCallbacks below.
-func HashStructure(s interface{}, cb HashCallback, ignoredKeys []string) (interface{}, error) {
-	s, err := copystructure.Copy(s)
+func HashStructure(s map[string]interface{}, cb HashCallback, ignoredKeys []string) (interface{}, error) {
+	scopy, err := copystructure.Copy(s)
```

Comment:
What's the reason for this copy? It assumes that you'd never want to modify the actual value coming in. Plus, we are already providing a copy in the first place.

Reply:
The copy was already there.  I wondered myself why we're copying twice.  The reason I introduced a new variable here was a side-effect of the signature change.

Reply:
Sure, it's already there -- the question is, what's the reason for it :-)

It seems like either we should figure out if we're copying in all cases already and remove this, or remove the places where we're already copying.

Reply:
Sounds good.  I'm thinking I'll remove the copying from HashStructure and FormatRequest/FormatResponse, instead doing it in Hash.  Rationale: HashStructure is lower level, maybe someday a caller will come along that doesn't need/want copying and this provides that option.  Meanwhile FormatRequest and FormatResponse are already doing a lot, and this detail feels like it should be Hash's responsibility.

Reply:
Sounds reasonable.